presto-main/src/main/java/com/facebook/presto/execution/executor/MultilevelSplitQueue.java

```diff
@@ -0,0 +1,322 @@
+                double ratio = levelScheduledTime[level] == 0 ? 0 : expectedScheduledTime / (1.0 * levelScheduledTime[level]);
+                if (selectedLevel == -1 || ratio > worstRatio) {
+                    worstRatio = ratio;
+                    selectedLevel = level;
+                }
+            }
+
+            expectedScheduledTime /= levelPriorityMultiplier;
+        }
+
+        if (selectedLevel == -1) {
+            return null;
+        }
+
+        PrioritizedSplitRunner result = levelWaitingSplits.get(selectedLevel).poll();
+        checkState(result != null, "pollSplit cannot return null");
+
+        return result;
+    }
+
+    /**
+     * Normalizes the scheduled time counters for empty levels.
+     *
+     * If only one level has waiting splits, we use it's current scheduled time value to compute the
+     * expected time for other levels.
+     *
+     * If multiple levels have waiting splits, the expected scheduled time is computed based on the
+     * scheduled time of the level that deviates most from the expected value
+     *
+     * @return expected scheduled time for level 0
+     */
+    @GuardedBy("lock")
+    private long updateLevelTimes()
+    {
+        long level0ExpectedTime = levelScheduledTime[0];
+        boolean updated;
+        do {
+            double currentMultiplier = levelPriorityMultiplier;
+            updated = false;
+            for (int level = 0; level < LEVEL_THRESHOLD_SECONDS.length; level++) {
+                currentMultiplier /= levelPriorityMultiplier;
+                long levelExpectedTime = (long) (level0ExpectedTime * currentMultiplier);
+
+                if (levelWaitingSplits.get(level).isEmpty()) {
+                    levelScheduledTime[level] = levelExpectedTime;
+                    continue;
+                }
+
+                if (levelScheduledTime[level] > levelExpectedTime) {
+                    level0ExpectedTime = levelScheduledTime[level] * (long) Math.pow(levelPriorityMultiplier, level);
+                    updated = true;
+                    break;
+                }
+            }
+        } while (updated && level0ExpectedTime != 0);
+
+        return level0ExpectedTime;
+    }
+
+    @GuardedBy("lock")
+    private PrioritizedSplitRunner pollFirstSplit()
+    {
+        for (PriorityQueue<PrioritizedSplitRunner> level : levelWaitingSplits) {
+            PrioritizedSplitRunner split = level.poll();
+            if (split != null) {
+                return split;
+            }
+        }
+
+        return null;
+    }
+
+    public Priority updatePriority(Priority oldPriority, long quantaNanos, long scheduledNanos)
+    {
+        int oldLevel = oldPriority.getLevel();
+        int newLevel = computeLevel(scheduledNanos);
+
+        // prevents a misbehaving task from unfairly penalizing all other tasks in its level
+        long levelContribution = Math.min(quantaNanos, LEVEL_CONTRIBUTION_CAP);
```

Comment:
Can we add something like?
```
// For example, if a read hung for several minutes, this is considered a bug or failing external dependency, so only penalize the task for a fixed portion.
```
Also it isn't clear how this would `penalizing all other tasks in its level`